wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+    }
+
+    // State C (RUNNING_PRIORITY = 0, PEND_INTERRUPT = 0b0000_0000)
+
+    loop {}
+}
+
+// INTERRUPT_PRIORITY[0] = 1
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    // State B (RUNNING_PRIORITY = 1, PEND_INTERRUPT = 0b0000_0000)
+}
+```
+
+Here setting the state of the first interrupt to *pending* causes the
+`INTERRUPT0` handler to preempt the `main` function.
+
+### Priorities
+
+``` rust
+#![no_std]
+
+use core::{cell::UnsafeCell, ptr};
+
+extern "C" {
+    static PEND_INTERRUPT: UnsafeCell<u8>;
+    static INTERRUPT_PRIORITY: [UnsafeCell<u8>; 8];
+}
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    unsafe {
+        // set the priority of `INTERRUPT1` to `2`
+        ptr::write_volatile(INTERRUPT_PRIORITY[1].get(), 2);
+        asm!("INTERRUPTS_ENABLE" : : : : "volatile");
+    }
+
+    // State A (RUNNING_PRIORITY = 0, PEND_INTERRUPT = 0b0000_0000)
+
+    unsafe {
+        // Set the state of `INTERRUPT0` to pending
+        ptr::write_volatile(PEND_INTERRUPT.get(), 1 << 0);
+    }
+
+    // State E (RUNNING_PRIORITY = 0, PEND_INTERRUPT = 0b0000_0000)
+
+    loop {}
+}
+
+// INTERRUPT_PRIORITY[0] = 1
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    // State B (RUNNING_PRIORITY = 1, PEND_INTERRUPT = 0b0000_0000)
+
+    unsafe {
+        // Set the state of `INTERRUPT1` to pending
+        ptr::write_volatile(PEND_INTERRUPT.get(), 1 << 1);
+    }
+
+    // State D (RUNNING_PRIORITY = 1, PEND_INTERRUPT = 0b0000_0000)
+}
+
+// INTERRUPT_PRIORITY[1] = 2
+#[no_mangle]
+unsafe fn INTERRUPT1() {
+    // State C (RUNNING_PRIORITY = 2, PEND_INTERRUPT = 0b0000_0000)
+}
+```
+
+Higher priority interrupt handlers can preempt lower priority ones.
+
+### Withholding an interrupt
+
+``` rust
+#![no_std]
+
+use core::{cell::UnsafeCell, ptr};
+
+extern "C" {
+    static PEND_INTERRUPT: UnsafeCell<u8>;
+    static INTERRUPT_PRIORITY: [UnsafeCell<u8>; 8];
+}
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    unsafe {
+        // set the priority of `INTERRUPT0` to `2`
+        ptr::write_volatile(INTERRUPT_PRIORITY[0].get(), 2);
+        asm!("INTERRUPTS_ENABLE" : : : : "volatile");
+    }
+
+    // State A (RUNNING_PRIORITY = 2, PEND_INTERRUPT = 0b0000_0000)
+
+    unsafe {
+        // Set the state of INTERRUPT0 to pending
+        ptr::write_volatile(PEND_INTERRUPT.get(), 1 << 0);
+    }
+
+    // State E (RUNNING_PRIORITY = 2, PEND_INTERRUPT = 0b0000_0000)
+
+    loop {}
+}
+
+// INTERRUPT_PRIORITY[0] = 2
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    // State B (RUNNING_PRIORITY = 2, PEND_INTERRUPT = 0b0000_0000)
+
+    unsafe {
+        // Set the state of INTERRUPT1 to pending
+        ptr::write_volatile(PEND_INTERRUPT.get(), 1 << 1);
+    }
+
+    // State C (RUNNING_PRIORITY = 2, PEND_INTERRUPT = 0b0000_0010)
+    //                                                          ^
+}
+
+// INTERRUPT_PRIORITY[1] = 1
+#[no_mangle]
+unsafe fn INTERRUPT1() {
+    // State D (RUNNING_PRIORITY = 1, PEND_INTERRUPT = 0b0000_0000)
+}
+```
```

Comment:
From reading the examples, it might be clearer to use `VolatileCell` instead of `UnsafeCell` for the registers.